### ISSUE TITLE
MPI version hangs

### ISSUE THREAD

**Issue by poquirion** [Cibiv/IQ-TREE#129]
I have compiled the code for mpi and openmp with the instructions form here:
http://www.iqtree.org/doc/Compilation-Guide#compiling-mpi-version

commit: 929aa69f1f3a86879c9486a519cc513bed073ca9 (HEAD -> master, tag: 2.0-rc1, origin/master, origin/HEAD):

gcc = 7.3.0
openmpi = 3.2.1
eigen = 3.3.5

I am testing it on a small dataset, it seems to run the analysis properly but then the mpi processes nevers returns and  hangs forever after having logged the following:

 ```
$ mpirun -np  4 iqtree-mpi  -redo -nt 1    -ninit 2 -n 2 -me 0.05  -s results/subsampled_alignment.fasta -m GTR
IQ-TREE MPI multicore version 1.6.12 for Linux 64-bit built Apr 24 2020
Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,
Heiko Schmidt, Dominik Schrempf, Michael Woodhams.

Host:    blg8547.int.ets1.calculquebec.ca (AVX512, FMA3, 92 GB RAM)
Command: iqtree-mpi -redo -nt 1 -ninit 2 -n 2 -me 0.05 -s results/subsampled_alignment.fasta -m GTR
Seed:    46616 (Using SPRNG - Scalable Parallel Random Number Generator)
Time:    Fri Apr 24 14:53:48 2020
Kernel:  AVX+FMA - 1 threads (40 CPU cores detected)
MPI:     4 processes

HINT: Use -nt option to specify number of threads because your CPU has 40 cores!
HINT: -nt AUTO will automatically determine the best number of threads to use.

[...]
--------------------------------------------------------------------
|                    FINALIZING TREE SEARCH                        |
--------------------------------------------------------------------
Performs final model parameters optimization
Estimate model parameters (epsilon = 0.050)
1. Initial log-likelihood: -41319.665
Optimal log-likelihood: -41319.651
Rate parameters:  A-C: 0.34331  A-G: 2.40006  A-T: 0.42603  C-G: 0.56344  C-T: 5.47531  G-T: 1.00000
Base frequencies:  A: 0.299  C: 0.184  G: 0.196  T: 0.321
Parameters optimization took 1 rounds (0.005 sec)
BEST SCORE FOUND : -41319.651
Total tree length: 0.003

Total number of iterations: 7
CPU time used for tree search: 0.025 sec (0h:0m:0s)
Wall-clock time used for tree search: 0.055 sec (0h:0m:0s)
Total CPU time used: 0.089 sec (0h:0m:0s)
Total wall-clock time used: 0.185 sec (0h:0m:0s)

Analysis results written to: 
  IQ-TREE report:                results/subsampled_alignment.fasta.iqtree
  Maximum-likelihood tree:       results/subsampled_alignment.fasta.treefile
  Likelihood distances:          results/subsampled_alignment.fasta.mldist
  Screen log file:               results/subsampled_alignment.fasta.log

Date and Time: Fri Apr 24 14:53:48 2020
```

I am suspecting the fact that the dataset is so small is the problem...

**Comment by poquirion** [comment]
I tested it on a large dataset that gives enough time for all the mpi process initialise  and I get the same result. It hangs once all has been computed.

**Comment by poquirion** [comment]
I have been debugging a bit. Here is what I have found:

The master calls MPIHelper::finalize() without ever sending the sendStopMessage() after the main loop of the IQTREE algo has ran.

it seems that this block
https://github.com/Cibiv/IQ-TREE/blob/929aa69f1f3a86879c9486a519cc513bed073ca9/tree/iqtree.cpp#L2244
should be 
```
    if (stop_rule.meetStopCondition(stop_rule.getCurIt(), cur_correlation)) {
        cout << "--------------------------------------------------------------------" << endl;
        cout << "|               OPTIMIZING CANDIDATE TREE SET                      |" << endl;
        cout << "--------------------------------------------------------------------" << endl;
    }

```
While the     `optimization_looped = true;` should be set at the end of the loop here:
https://github.com/Cibiv/IQ-TREE/blob/929aa69f1f3a86879c9486a519cc513bed073ca9/tree/iqtree.cpp#L2371

But then I am not even sure why the `optimization_looped` variable exist and since if the condition is not me, the loop will go on forever and the process will not complete... I would just remove the optimization_looped and the if condition here:
https://github.com/Cibiv/IQ-TREE/blob/929aa69f1f3a86879c9486a519cc513bed073ca9/tree/iqtree.cpp#L2382

But in any case, it seem that moving the optimization_looped call solves the hanging worker problems.

**Comment by poquirion** [comment]
Ok, it seems that the  master process never gets in the main loop but the worker do, hence  `optimization_looped` is always false for the master and the `stop` signal is never sent to the workers.

**Comment by bqminh** [comment]
Thanks for reporting this and the information is very helpful! Indeed that is the case. This is due to -n option. Without this option everything works. I'll fix in the next commit. 
However, I just want to note that the MPI version only scales with higer number of iterations (-n). With -n 2 there is no gain in speedup. Therefore, even after the fix, I don't recommend to use the MPI version when you set this low number of iterations.

Minh

**Comment by bqminh** [comment]
Fixed now in commit 8dc02096cc
